sql: minor fixes for opclasses in trigram indexes

mgartner opened the pull request:
#### sql: err when an opclass is provided for a non-inverted index column

Fixes #84512

Release justification: This is a minor fix for new features.

Release note: None

#### sql: do not show opclasses for non-inverted index columns

Fixes #86614

Release justification: This is a minor bug fix for a new feature.

Release note: None

mgartner committed:
sql: err when an opclass is provided for a non-inverted index column

Fixes #84512

Release justification: This is a minor fix for new features.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87206)
<!-- Reviewable:end -->

mgartner committed:
sql: do not show opclasses for non-inverted index columns

Fixes #86614

Release justification: This is a minor bug fix for a new feature.

Release note: None

ajwerner reviewed (commented):
We should probably add descriptor validation to disallow an opclass in an index key too. I'll file a follow-up issue. 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-:-NApCLu-BsW7VWjlPrBQ:b66engs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, @jordanlewis, and @mgartner)

___
*[`pkg/sql/create_index.go` line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApBuIP-RGJFuxUCDKK:-NApBuIP-RGJFuxUCDKL:b9k51ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe6f46d7e12a937076d06b2a6aa5dc9e9b4b3843/pkg/sql/create_index.go#L327)):*
> ```Go
> 		}
> 		if colDef.OpClass != "" && (i < len(columns)-1 || !inverted) {
> 			return pgerror.New(pgcode.UndefinedObject,
> ```

I don't know about `UndefinedObject` here, how'd you choose it?
___
*[`pkg/sql/logictest/testdata/logic_test/trigram_indexes` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApC9HhA9K0J-GC30a4:-NApC9HhA9K0J-GC30a5:b-svlb4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6db072ee475f480c40780a6c3f61bf8011576a3f/pkg/sql/logictest/testdata/logic_test/trigram_indexes#L15)):*
> ```
> statement error operator classes are only allowed for the last column of an inverted index
> CREATE INVERTED INDEX ON a (a gin_trgm_ops, t gin_trgm_ops)
> ```

may I request `pgcode` checking? In general I think it's a mistake when we add new logic test cases which assert errors but not pgcodes



_Code quote:_
```
statement error data type text has no default operator class for access method \"gin\"
CREATE INVERTED INDEX ON a(t)

statement error data type text has no default operator class for access method \"gin\"
CREATE INDEX ON a USING GIN(t)

statement error operator classes are only allowed for the last column of an inverted index
CREATE INDEX ON a (t gin_trgm_ops)

statement error operator classes are only allowed for the last column of an inverted index
CREATE INVERTED INDEX ON a (a gin_trgm_ops, t gin_trgm_ops)
```


<!-- Sent from Reviewable.io -->

jordanlewis commented on pkg/sql/catalog/catformat/index.go:
```diff
@@ -220,7 +220,7 @@ func FormatIndexElements(
 			f.FormatNameP(&index.KeyColumnNames[i])
 		}
-		if index.Type == descpb.IndexDescriptor_INVERTED && len(index.InvertedColumnKinds) > 0 {
+		if i == n-1 && index.Type == descpb.IndexDescriptor_INVERTED && len(index.InvertedColumnKinds) > 0 {
```
Comment? Or use an API? Isn't there a "get inverted index column api"?

jordanlewis reviewed (approved):
LGTM

ajwerner commented:
https://github.com/cockroachdb/cockroach/issues/87216

mgartner commented:
> We should probably add descriptor validation to disallow an opclass in an index key too. I'll file a follow-up issue.

Note that the validation will be pretty simple - there is no place for opclasses for non-inverted columns in the descriptor. So that validation would just be `len(InvertedColumnKinds) == 0 || (IsInverted && len(InvertedColumnKinds) == 1))`.

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @DrewKimball, and @jordanlewis)

___
*[`pkg/sql/create_index.go` line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApBuIP-RGJFuxUCDKK:-NAufeUM3D3Bl0chTdno:bows4yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe6f46d7e12a937076d06b2a6aa5dc9e9b4b3843/pkg/sql/create_index.go#L327)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't know about `UndefinedObject` here, how'd you choose it?
</blockquote></details>

This is what PG uses:

```
marcus=# create table t (t text);
CREATE TABLE
marcus=# create index on t (t gin_trgm_ops);
ERROR:  42704: operator class "gin_trgm_ops" does not exist for access method "btree"
LOCATION:  ResolveOpClass, indexcmds.c:2115
```

I'd prefer a `InvalidIndexDefinition` but none exists. Open to suggestions though.
___
*[`pkg/sql/catalog/catformat/index.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApBft9SkgrQ_eYR-Dz-r2-223:-NAuhckL9EoX5gJfoA_X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6db072ee475f480c40780a6c3f61bf8011576a3f/pkg/sql/catalog/catformat/index.go#L223)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Comment? Or use an API? Isn't there a "get inverted index column api"?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/trigram_indexes` line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApC9HhA9K0J-GC30a4:-NAugjCw33Kz9QrAHUHD:b-yuxowa) ([raw file](https://github.com/cockroachdb/cockroach/blob/6db072ee475f480c40780a6c3f61bf8011576a3f/pkg/sql/logictest/testdata/logic_test/trigram_indexes#L15)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

may I request `pgcode` checking? In general I think it's a mistake when we add new logic test cases which assert errors but not pgcodes
</blockquote></details>

Thanks for calling this out - I agree. :+1: Will be more diligent with this going forward.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/create_index.go` line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87206#-NApBuIP-RGJFuxUCDKK:-NAuiMBy7BcsDydz0MoA:bquvkv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe6f46d7e12a937076d06b2a6aa5dc9e9b4b3843/pkg/sql/create_index.go#L327)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is what PG uses:

```
marcus=# create table t (t text);
CREATE TABLE
marcus=# create index on t (t gin_trgm_ops);
ERROR:  42704: operator class "gin_trgm_ops" does not exist for access method "btree"
LOCATION:  ResolveOpClass, indexcmds.c:2115
```

I'd prefer a `InvalidIndexDefinition` but none exists. Open to suggestions though.
</blockquote></details>

They also use DatatypeMismatch 42804 in some cases:

```
marcus=# create index on t using gin (a gin_trgm_ops);
ERROR:  42804: operator class "gin_trgm_ops" does not accept data type integer
LOCATION:  ResolveOpClass, indexcmds.c:2137
```


<!-- Sent from Reviewable.io -->

ajwerner commented:
`DatatypeMismatch` feels nice to me

mgartner commented:
TFTRs!

bors r+
